cmd/encryption-v1.go

```diff
@@ -331,14 +202,14 @@ func rotateKey(oldKey []byte, newKey []byte, bucket, object string, metadata map
 		return errors.New("failed to seal object encryption key") // if this happens there's a bug in the code (may panic ?)
 	}
 
-	metadata[ServerSideEncryptionIV] = base64.StdEncoding.EncodeToString(iv[:])
-	metadata[ServerSideEncryptionSealAlgorithm] = SSESealAlgorithmDareV2HmacSha256
-	metadata[ServerSideEncryptionSealedKey] = base64.StdEncoding.EncodeToString(sealedKeyW.Bytes())
+	metadata[crypto.SSEIV] = base64.StdEncoding.EncodeToString(iv[:])
+	metadata[crypto.SSESealAlgorithm] = SSESealAlgorithmDareV2HmacSha256
+	metadata[crypto.SSECSealedKey] = base64.StdEncoding.EncodeToString(sealedKeyW.Bytes())
 	return nil
 }
 
-func newEncryptMetadata(key []byte, bucket, object string, metadata map[string]string) ([]byte, error) {
-	delete(metadata, SSECustomerKey) // make sure we do not save the key by accident
+func newEncryptMetadata(key []byte, bucket, object string, metadata map[string]string, sseS3 bool) ([]byte, error) {
+	delete(metadata, crypto.SSECKey) // make sure we do not save the key by accident
 	// See crypto/doc.go for detailed description
 	nonce := make([]byte, 32+SSEIVSize) // generate random values for key derivation
```

Comment:
You have to replace the following lines with:
```
var clientKey, iv [32]byte
var sealedKey crypto.SealedKey
if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
 	logger.CriticalIf(context.Background(), errors.New("The system PRNG is out of entropy")
}
copy(clientKey[:], key)
objectKey := crypto.GenerateKey(clientKey, rand.Reader)
if sseS3 {
sealedKey = objectKey.Seal(clientKey, iv, "SSE-S3", bucket, object)
metadata = crypto.S3.CreateMetadata(metadata, sealedKey)
} else {
sealedKey = objectKey.Seal(clientKey, iv, "SSE-C", bucket, object)
metadata = crypto.SSEC.CreateMetadata(metadata, sealedKey)
}
return objectKey
```

Reply:
Also don't worry about re-usability too much here - that's a separate task I'll take on. This PR is just for vault KMS integration...

Reply:
@aead, the crypto.S3.CreateMetadata() is overlaying the Vault sealed key that we need to store in the metadata by further sealing the sealed key from vault.

Reply:
@poornas Are you sure? I think the problem is that the decryption uses the wrong domain when decrypting. `CreateMetadata` only adds values to the metadata (if it's not nil). If you pass the created KMS key and key-ID to `newEncryptMetadata` it should work

Reply:
The problem with the current code is that it uses the `SSE-C` domain even for `SSE-S3`.

Reply:
@aead, to answer your previous question, tested it  with above code and it causes decryption failure because the sealed key is overlaid. So, will you be changing Seal/Unseal methods of key.go